### DISCUSSION TITLE
Declaration of Psr\Log\NullLogger must be compatible with Psr\Log\AbstractLogger

### DISCUSSION THREAD

**Discussion by torbentschechne** [mpdf/mpdf#1956] [General]
Hello,

I use mPDF in my code to generate PDFs. The server runs PHP 8.0 and suddenly I get this error:

```
PHP Fatal error: Declaration of Psr\Log\NullLogger::log($level, Stringable|string $message, array $context = []): void must be compatible with Psr\Log\AbstractLogger::log($level, $message, array $context = []) in /files/wp-content/plugins/pdf-creator/includes/vendor/psr/log/src/NullLogger.php on line 26
```

The weird thing is, the files look okay. The `NullLogger` and the `LoggerTrait` have the same signature.

LoggerTrait (which feeds the AbstractLogger):

```
abstract public function log($level, string|\Stringable $message, array $context = []): void;
```

NullLogger:

```
    public function log($level, string|\Stringable $message, array $context = []): void
    {
        // noop
    }
```

I updated from mPDF 8.1.* to mPDF 8.2.2. I switched from `psr/log` 2 to `psr/log` 3. But still when using 8.2.2 and/or `psr/log` in Version 2 or 3, I get the error mentioned above. So I wonder, where is this error/the wrong signature coming from? 

I also checked the server, but nothing regarding this is installed in my php installation:

```
sudo dpkg --get-selections | grep php
php-common					install
php8.0-bcmath					install
php8.0-cli					install
php8.0-common					install
php8.0-curl					install
php8.0-fpm					install
php8.0-gd					install
php8.0-igbinary					install
php8.0-imagick					install
php8.0-intl					install
php8.0-mbstring					install
php8.0-mysql					install
php8.0-opcache					install
php8.0-readline					install
php8.0-redis					install
php8.0-soap					install
php8.0-xml					install
php8.0-zip					install
```

Can anybody tell me what's going on and how I can fix it?

**Comment by finwe** [comment]
Post your `composer info mpdf/*` and `composer info psr/*` output.

**Reply by torbentschechne** [reply]
```
mpdf/mpdf                  v8.2.2 PHP library generating PDF files from UTF-8 encoded HTML
mpdf/psr-http-message-shim v2.0.1 Shim to allow support of different psr/message versions.
mpdf/psr-log-aware-trait   v3.0.0 Trait to allow support of different psr/log versions.
```

**Reply by torbentschechne** [reply]
I removed the `vendor` and `composer.lock` and just ran `composer install` with this composer.json

```
{
    "require": {
        "mpdf/mpdf": "~8.2"
    }
}

```

Still, I get this error. Also I did `composer clear-cache` and `composer dump-autoload` - no success.

**Reply by finwe** [reply]
This seems OK. Can you try deleting the vendor directory and running `composer install`?

**Reply by finwe** [reply]
Are you running fpm? Can you try reloading it or restarting the server? There might be some cache.

**Reply by torbentschechne** [reply]
Yes, fpm is running. I restartet fpm, nginx and the whole server. No success :/

**Reply by finwe** [reply]
I'm out of ideas.

**Reply by torbentschechne** [reply]
Whats really odd, I did this as explained here: 

https://stackoverflow.com/a/76243472/1124269

So my composer.json looks like this:

```
{
    "require": {
        "mpdf/mpdf": "~8.2",
        "psr/log": "^1.1"
    }
}
```

And now it works...

**Reply by johnvanham** [reply]
> Whats really odd, I did this as explained here:
> 
> https://stackoverflow.com/a/76243472/1124269
> 
> So my composer.json looks like this:
> 
> ```
> {
>     "require": {
>         "mpdf/mpdf": "~8.2",
>         "psr/log": "^1.1"
>     }
> }
> ```
> 
> And now it works...

Just had to do the same on our project, using PHP 8.1.. was fine before a recent update although I don't know which as this project has been in production for a while without any new deployments until recently.

**Comment by khanhvo** [comment]
Hi there,

I am facing this issue on both PHP 8.0 and PHP 8.2. Any advice, please?

My packages:
Package operations: 8 installs, 0 updates, 0 removals
  - Downloading paragonie/random_compat (v9.99.100)
  - Installing setasign/fpdi (v2.6.0): Extracting archive
  - Installing psr/log (3.0.0): Extracting archive
  - Installing psr/http-message (2.0): Extracting archive
  - Installing paragonie/random_compat (v9.99.100): Extracting archive
  - Installing myclabs/deep-copy (1.11.1): Extracting archive
  - Installing mpdf/psr-log-aware-trait (v3.0.0): Extracting archive
  - Installing mpdf/psr-http-message-shim (v2.0.1): Extracting archive
  - Installing mpdf/mpdf (v8.2.3): Extracting archive

**Comment by carrguerr** [comment]
Removed string|\Stringable from /vender/psr/log/src/*php files and works now
